.github/workflows/test-and-build.yml

```diff
@@ -0,0 +1,194 @@
+    needs: node_setup
+    runs-on: ubuntu-latest
+    env:
+      CI: true
+    steps:
+      - uses: actions/checkout@v4
+
+      - uses: actions/setup-node@v4
+        with: { node-version: "${{ needs.node_setup.outputs.node_version }}" }
+
+      - uses: actions/cache@v4
+        env:
+          cache-name: cache-node-modules
+        with:
+          path: ~/.npm
+          key: ${{ runner.os }}-build-${{ env.cache-name }}-${{ hashFiles('**/package-lock.json') }}
+          restore-keys: |
+            ${{ runner.os }}-build-${{ env.cache-name }}-
+            ${{ runner.os }}-build-
+            ${{ runner.os }}-
+
+      - name: Install Dependencies
+        run: npm ci
+
+      - name: Test
+        run: npm run test:ci -- --coverage
+
+      - name: Coveralls Report
+        uses: coverallsapp/github-action@master
+        with:
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+
+  publish:
+    name: Publish image to Docker registry
+    needs: [lint, typecheck, test]
+
+    env:
+      REGISTRY_HOST: ghcr.io
+      IMAGE_PUSH_ENABLED: ${{ secrets.NO_DOCKER_IMAGE == null }}
+      # This gets us a real SHA for the SHA tag, instead of an ethereal one. See:
+      # - https://github.com/docker/metadata-action?tab=readme-ov-file#environment-variables
+      # - https://github.com/docker/metadata-action/issues/206
+      DOCKER_METADATA_PR_HEAD_SHA: true
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+      packages: write
+
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Generate tags for Docker image
+        id: docker-image-tags
+        uses: docker/metadata-action@v5
+        with:
+          images: ${{ env.REGISTRY_HOST }}/${{ github.repository_owner }}/${{ github.event.repository.name }}
+          tags: |
+            type=semver,pattern={{major}}.{{minor}},priority=10
+            type=semver,pattern={{version}},priority=20
+            type=ref,event=branch,priority=30
+            type=sha,priority=40
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+
+      - name: Login to the Docker registry
+        uses: docker/login-action@v3
+        with:
+          registry: ${{ env.REGISTRY_HOST }}
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Cache Docker layers
+        uses: actions/cache@v4
+        with:
+          path: /tmp/.buildx-cache
+          key: ${{ runner.os }}-buildx-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-buildx-
+
+      - name: Build and push Docker image
+        uses: docker/build-push-action@v6
+        with:
+          context: .
+          file: ./Dockerfile
+          push: ${{ env.IMAGE_PUSH_ENABLED }}
+          tags: ${{ steps.docker-image-tags.outputs.tags }}
+          cache-from: type=local,src=/tmp/.buildx-cache
+          cache-to: type=local,dest=/tmp/.buildx-cache
```

Comment:
Minor: This will work, but I believe the cache will grow unbounded, until its 10Gb and then github will automatically delete it, then the process will start again. You might instead want to consider using a [registry cache](https://docs.docker.com/build/ci/github-actions/cache/#registry-cache) or the [github cache](https://docs.docker.com/build/ci/github-actions/cache/#github-cache), since they both provide better targeted cache management.

This work could be a follow up though, it becomes a problem. Just wanted to note those other cache sources, so you knew about them.